docs/Contributing.asciidoc

```diff
@@ -333,6 +333,42 @@ Authentication module is expected to return HASH:
 Authentication module is expected to create or update user entry in OpenQA database
 after user validation. See included modules for inspiration.
 
+== WebAPI Plugins (requires Mojolicious::Plugin::OpenAPI)
+
+During processing of `openqa.ini` all sections with prefix "WebAPIPlugin" are examined 
+for parameters `enabled` and `path`.
+(Tail of the section name is considered to be name of the Plugin).
+Unless parameter `enabled` is missing or set to `0` - the process will attempt to load plugin from 
+`WebAPIPlugin` folder within `path` configured. In particular following artefects will be loaded:
+* Json file with OpenAPI format describing routes under `/plugin` entry point.
+* Perl module.
+
+Parameters from the configration file secion will be passed to plugin's `init()` method.
+
+E.g. with section like below in the openqa.ini:
+[source]
+----
+[WebAPIPluginFoo]
+enabled=optional
+path=/etc/bar
+Msg=Test
+----
+The service will attempt to setup routes from json file with OpenAPI format 
+`/etc/bar/WebAPIPlugin/Foo.json` and load perl module: `/etc/bar/WebAPIPlugin/Foo.pm`.
+`Msg` parameter will be passed to plugin `register()` method.
+
+At the moment following restrictions apply:
+* package Mojolicious::Plugin::OpenAPI must be installed on the system.
```

Comment:
Do you really want to limit webui extensions to OpenAPI? Mojolicious has a perfectly fine extension mechanism that's extremely versatile and that you could just reuse with a custom namespace like `OpenQA::WebAPI::Extensions`.

Reply:
A good example might be the [Minion admin ui](https://github.com/mojolicious/minion/blob/master/lib/Mojolicious/Plugin/Minion/Admin.pm), which is a plugin we actually use in openQA too.

Reply:
In my view when somebody implements other routing mechanism - they will update this readme.
I don't have much preference how routing is imported - I just want a way to show my extension in OpenQA UI.

Reply:
In my understanding all routes in the Minion ui ar epredefined, so it is not good example? Or where should I look to?

Reply:
The routes get attached to a prefix route when the plugin is loaded, completely dynamic. Any Mojolicious application can load the plugin and attach it to any prefix.

Reply:
The routes the Minion UI plugin adds are defined directly in the source code of that plugin.

Seems like you want to create a plugin here which loads its routes dynamically from OpenAPI definitions. That's different, indeed.

Reply:
Not really, the Minion ui plugin could use the OpenAPI plugin too. There are no limits to what a Mojolicious plugin can do, it's just like the code in the `startup` sub of `OpenQA::WebAPI`.

Reply:
still "The routes the Minion UI plugin adds are defined directly in the source code", so what format do you suggest to use for abstract pluggable module from external project? We must pass/parse routes definition somehow and what is alternative to OpenAPI? (except of custom implementation of what OpenAPI does).

Reply:
Ah now I see what you meant by "The routes get attached to a prefix route when the plugin is loaded, completely dynamic", I've deleted my previous comment.

Reply:
Maybe it is the way to go without OpenAPI at all, at least at the moment. I must try it.